crucible-concurrency/src/Cruces/CrucesMain.hs

```diff
@@ -0,0 +1,167 @@
+import Data.Parameterized.Some (Some(Some))
+
+import qualified Lang.Crucible.CFG.Core as C
+import Lang.Crucible.CFG.Reg
+import Lang.Crucible.CFG.SSAConversion
+
+import Lang.Crucible.Syntax.Concrete
+import Lang.Crucible.Syntax.SExpr
+import Lang.Crucible.Syntax.ExprParse (printSyntaxError)
+import Lang.Crucible.Syntax.Atoms
+
+import Lang.Crucible.Analysis.Postdom
+import Lang.Crucible.Backend
+import Lang.Crucible.FunctionHandle
+import Lang.Crucible.Simulator
+
+import What4.Interface
+import What4.ProgramLoc
+import What4.FunctionName
+
+import qualified Crux
+import           Crux.Types
+
+import Data.Parameterized.Pair
+import qualified SimpleGetOpt as GetOpt
+import Text.Read (readMaybe)
+
+
+import Crucibles.Primitives
+import Crucibles.DPOR (DPOR)
+import Crucibles.ExploreTypes
+import Cruces.ExploreCrux
+
+data CrucesOptions = CrucesOptions
+  { noDpor :: Bool
+  , debug  :: Bool
+  , pedantic :: Bool
+  , maxPreemptions :: Int
+  , target :: FilePath
+  }
+
+defaultCrucesOptions :: CrucesOptions
+defaultCrucesOptions = CrucesOptions
+  { noDpor = False
+  , debug  = False
+  , pedantic = False
+  , maxPreemptions = 0
+  , target = ""
+  }
+
+cruciblesConfig :: Crux.Config CrucesOptions
+cruciblesConfig = Crux.Config
+  { Crux.cfgFile =  pure defaultCrucesOptions
+  , Crux.cfgEnv  = []
+  , Crux.cfgCmdLineFlag =
+      [ GetOpt.Option [] ["no-dpor"]
+        "Enumerate all paths naively"
+        (GetOpt.NoArg (\opts -> Right opts { noDpor = True }))
+      , GetOpt.Option [] ["pedantic"]
+        "Pedantic signal behavior (spurious wakeups)"
+        (GetOpt.NoArg (\opts -> Right opts { pedantic = True }))
+      , GetOpt.Option [] ["max-preemptions"]
+        "Maximum number of preemptive context switches"
+        (GetOpt.ReqArg "NUM" (\x opts ->
+                                case readMaybe x of
+                                  Just i  -> Right opts { maxPreemptions = i }
+                                  Nothing -> Left "--max-preemptions requires an integer"))
+      ]
+  }
+
+findMain :: FunctionName -> [ACFG] -> FnVal sym Ctx.EmptyCtx C.UnitType
+findMain mainName cs =
+  case find (isFn mainName) cs of
+    Just (ACFG Ctx.Empty C.UnitRepr m) ->
+      HandleFnVal (cfgHandle m)
+    _ ->
+      undefined
+  where
+    isFn x (ACFG _ _ g) = handleName (cfgHandle g) == x
+
+run :: (Crux.Logs) => (Crux.CruxOptions, CrucesOptions) -> IO ()
+run (cruxOpts, opts) =
+  do let ?dpor = not (noDpor opts)
+     let ?bound = maxPreemptions opts
+     let [fn] = Crux.inputFiles cruxOpts
+     -- Some ng <- newIONonceGenerator
+     ha      <- newHandleAllocator
+     theInput <- readFile fn
+     case MP.parse (skipWhitespace *> many (sexp atom) <* eof) fn (pack theInput) of
+       Left err ->
+         do putStrLn $ errorBundlePretty err
+            exitFailure
+       Right v ->
+         withIONonceGenerator $ \nonceGen ->
+           -- withYicesOnlineBackend FloatRealRepr nonceGen ProduceUnsatCores noFeatures $ \sym -> do
```

Comment:
Ditto. I imagine it inherits support for all of the standard Crux provers now. Is that right?